### PR TITLE
En 7014 prune data trie on remove account

### PR DESCRIPTION
**Pull request by BeniaminDrasovean**
Implemented a func so we can get all hashes from a certain trie.
When an account is removed, mark all the data trie hashes for eviction.

### CODE REVIEW

File: data/state/accountsDB_test.go [hunk 1]
```diff
@@ -1270,3 +1270,72 @@ func TestAccountsDB_MainTrieAutomaticallyMarksCodeUpdatesForEviction(t *testing.
 	hashesForEviction = ewl.Cache[string(rootHash2)]
 	assert.Equal(t, 3, len(hashesForEviction))
 }
+
+func TestAccountsDB_RemoveAccountSetsObsoleteHashes(t *testing.T) {
+	t.Parallel()
+
+	marshalizer := &mock.MarshalizerMock{}
+	hsh := mock.HasherMock{}
```
Comment (bogdan-rosianu) [inline data/state/accountsDB_test.go:1278]: add some empty spaces here so the test can be more readable

File: data/state/accountsDB_test.go [hunk 2]
```diff
@@ -1270,3 +1270,72 @@ func TestAccountsDB_MainTrieAutomaticallyMarksCodeUpdatesForEviction(t *testing.
...
+	_, _ = adb.Commit()
+
+	acc, _ = adb.LoadAccount(addr)
+	userAcc = acc.(state.UserAccountHandler)
+	userAcc.SetCode([]byte("code"))
+	snapshot := adb.JournalLen()
+
+	err := adb.RemoveAccount(addr)
+	assert.Nil(t, err)
+	assert.Equal(t, 1, len(adb.GetObsoleteHashes()))
```
Comment (bogdan-rosianu) [inline data/state/accountsDB_test.go:1301]: maybe assert also the equality, not just length 1

File: data/state/accountsDB_test.go [hunk 3]
```diff
@@ -1270,3 +1270,72 @@ func TestAccountsDB_MainTrieAutomaticallyMarksCodeUpdatesForEviction(t *testing.
...
+func TestAccountsDB_RemoveAccountMarksObsoleteHashesForEviction(t *testing.T) {
+	t.Parallel()
+
+	marshalizer := &mock.MarshalizerMock{}
+	hsh := mock.HasherMock{}
+	db := mock.NewMemDbMock()
+	ewl, _ := mock.NewEvictionWaitingList(100, mock.NewMemDbMock(), marshalizer)
+	storageManager, _ := trie.NewTrieStorageManager(db, marshalizer, hsh, config.DBConfig{}, ewl, config.TrieStorageManagerConfig{})
+	maxTrieLevelInMemory := uint(5)
+	tr, _ := trie.NewTrie(storageManager, marshalizer, hsh, maxTrieLevelInMemory)
```
Comment (bogdan-rosianu) [inline data/state/accountsDB_test.go:1317]: add some empty lines and you can move `maxTrieLevelInMemory` to the top of the test

File: data/state/accountsDB_test.go [hunk 4]
```diff
@@ -1270,3 +1270,72 @@ func TestAccountsDB_MainTrieAutomaticallyMarksCodeUpdatesForEviction(t *testing.
...
+	addr1 := make([]byte, 32)
+	addr1[0] = 1
+	acc, _ = adb.LoadAccount(addr1)
+	_ = adb.SaveAccount(acc)
+
+	rootHash, _ := adb.Commit()
+
+	err := adb.RemoveAccount(addr)
+	assert.Nil(t, err)
+	assert.Equal(t, 1, len(adb.GetObsoleteHashes()))
```
Comment (bogdan-rosianu) [inline data/state/accountsDB_test.go:1335]: assert also the equality, not just length 1

File: data/trie/branchNode.go [hunk 5]
```diff
@@ -729,3 +729,34 @@ func (bn *branchNode) getAllLeaves(leaves map[string][]byte, key []byte, db data
 
 	return nil
 }
+
+func (bn *branchNode) getAllHashes(db data.DBWriteCacher) ([][]byte, error) {
+	err := bn.isEmptyOrNil()
+	if err != nil {
+		return nil, fmt.Errorf("getAllHashes error %w", err)
```
Comment (bogdan-rosianu) [inline data/trie/branchNode.go:736]: `"getAllHashes error: %w"`

File: data/trie/branchNode_test.go [hunk 6]
```diff
@@ -1337,3 +1337,24 @@ func TestBranchNode_getDirtyHashesFromCleanNode(t *testing.T) {
...
 	assert.Equal(t, 0, len(dirtyHashes))
 }
+
+func TestBranchNode_getAllHashes(t *testing.T) {
+	t.Parallel()
+
+	bn, _ := getBnAndCollapsedBn(getTestMarshAndHasher())
+	hashes, err := bn.getAllHashes(mock.NewMemDbMock())
+	assert.Nil(t, err)
+	assert.Equal(t, 4, len(hashes))
```
Comment (bogdan-rosianu) [inline data/trie/branchNode_test.go:1347]: why 4 ?

File: data/trie/extensionNode.go [hunk 7]
```diff
@@ -592,3 +592,24 @@ func (en *extensionNode) getAllLeaves(leaves map[string][]byte, key []byte, db d
 
 	return nil
 }
+
+func (en *extensionNode) getAllHashes(db data.DBWriteCacher) ([][]byte, error) {
+	err := en.isEmptyOrNil()
+	if err != nil {
+		return nil, fmt.Errorf("getAllHashes error %w", err)
```
Comment (bogdan-rosianu) [inline data/trie/extensionNode.go:599]: `"getAllHashes error: %w"`

File: data/trie/extensionNode_test.go [hunk 8]
```diff
@@ -1027,3 +1027,24 @@ func TestExtensionNode_getDirtyHashesFromCleanNode(t *testing.T) {
...
 	assert.Equal(t, 0, len(dirtyHashes))
 }
+
+func TestExtensionNode_getAllHashes(t *testing.T) {
+	t.Parallel()
+
+	en, _ := getEnAndCollapsedEn()
+	hashes, err := en.getAllHashes(mock.NewMemDbMock())
+	assert.Nil(t, err)
+	assert.Equal(t, 5, len(hashes))
```
Comment (bogdan-rosianu) [inline data/trie/extensionNode_test.go:1037]: why 5 ?

File: data/trie/leafNode.go [hunk 9]
```diff
@@ -354,3 +354,12 @@ func (ln *leafNode) getAllLeaves(leaves map[string][]byte, key []byte, _ data.DB
 	leaves[string(nodeKey)] = ln.Value
 	return nil
 }
+
+func (ln *leafNode) getAllHashes(_ data.DBWriteCacher) ([][]byte, error) {
+	err := ln.isEmptyOrNil()
+	if err != nil {
+		return nil, fmt.Errorf("getAllHashes error %w", err)
```
Comment (bogdan-rosianu) [inline data/trie/leafNode.go:361]: `"getAllHashes error: %w"`

File: integrationTests/state/stateTrie/stateTrie_test.go [hunk 10]
```diff
@@ -1838,3 +1839,140 @@ func checkCodeConsistency(
...
+	removedAccounts map[int]struct{},
+	dataTriesRootHashes [][]byte,
+) {
+	for i, rootHash := range dataTriesRootHashes {
+		_, ok := removedAccounts[i]
+		if ok {
+			err := adb.RecreateTrie(rootHash)
+			assert.NotNil(t, err)
+		} else {
+			err := adb.RecreateTrie(rootHash)
```
Comment (bogdan-rosianu) [inline integrationTests/state/stateTrie/stateTrie_test.go:1974]: does this `RecreateTrie` affect the state of the object in the future uses? If yes, make a copy of it before checking that recreate trie does not error

File: data/state/errors.go [hunk 11]
```diff
@@ -109,3 +109,9 @@ var ErrInvalidErdAddress = errors.New("invalid ERD address")
 
 // ErrInvalidPubkeyConverterType signals that the provided pubkey converter type is invalid
 var ErrInvalidPubkeyConverterType = errors.New("invalid pubkey converter type")
+
+// ErrNilHashesMap signals that the provided hashes map is nil
+var ErrNilHashesMap = errors.New("nil hashes map")
```
Comment (sasurobert) [inline data/state/errors.go:114]: ErrNilMapOfHashes

File: data/state/journalEntries.go [hunk 12]
```diff
@@ -209,3 +209,37 @@ func (jedtu *journalEntryDataTrieUpdates) Revert() (AccountHandler, error) {
...
+	return &journalEntryDataTrieRemove{
+		rootHash:               rootHash,
+		obsoleteDataTrieHashes: obsoleteDataTrieHashes,
+	}, nil
+}
+
+// Revert applies undo operation
+func (jedtr *journalEntryDataTrieRemove) Revert() (AccountHandler, error) {
+	delete(jedtr.obsoleteDataTrieHashes, string(jedtr.rootHash))
+
```
Comment (sasurobert) [inline data/state/journalEntries.go:238]: You may delete the whole map here. After revert there is no use for this map

File: data/state/accountsDB_test.go (hunk 1 above)
Comment (BeniaminDrasovean) [reply data/state/accountsDB_test.go:1278]: Done.

File: data/state/accountsDB_test.go (hunk 2 above)
Comment (BeniaminDrasovean) [reply data/state/accountsDB_test.go:1301]: Done.

File: data/state/accountsDB_test.go (hunk 3 above)
Comment (BeniaminDrasovean) [reply data/state/accountsDB_test.go:1317]: Done.

File: data/state/accountsDB_test.go (hunk 4 above)
Comment (BeniaminDrasovean) [reply data/state/accountsDB_test.go:1335]: Done.

File: data/trie/branchNode.go (hunk 5 above)
Comment (BeniaminDrasovean) [reply data/trie/branchNode.go:736]: Done.

File: data/trie/branchNode_test.go (hunk 6 above)
Comment (BeniaminDrasovean) [reply data/trie/branchNode_test.go:1347]: Because there are 4 nodes in the trie.

File: data/trie/extensionNode.go (hunk 7 above)
Comment (BeniaminDrasovean) [reply data/trie/extensionNode.go:599]: Done.

File: data/trie/extensionNode_test.go (hunk 8 above)
Comment (BeniaminDrasovean) [reply data/trie/extensionNode_test.go:1037]: Because there are 5 trie nodes.

File: data/trie/leafNode.go (hunk 9 above)
Comment (BeniaminDrasovean) [reply data/trie/leafNode.go:361]: Done.

File: integrationTests/state/stateTrie/stateTrie_test.go (hunk 10 above)
Comment (BeniaminDrasovean) [reply integrationTests/state/stateTrie/stateTrie_test.go:1992]: There is no need for a copy here.

File: data/state/errors.go (hunk 11 above)
Comment (BeniaminDrasovean) [reply data/state/errors.go:114]: Done.

File: data/state/journalEntries.go (hunk 12 above)
Comment (BeniaminDrasovean) [reply data/state/journalEntries.go:238]: If there are more accounts removed in a block, but only one account removal is reverted, this map will hold the obsoleteDataTrieHashes for the removal of the other data tries.

**Pull request merged**